File: src/main/java/com/nobility/downloader/updates/UpdateManager.java
package com.nobility.downloader.updates;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.nobility.downloader.Model;
import com.nobility.downloader.settings.Defaults;
import javafx.application.Platform;

import javax.net.ssl.HttpsURLConnection;
import javax.net.ssl.SSLContext;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.URL;
import java.security.SecureRandom;
import java.security.cert.X509Certificate;

public class UpdateManager {

    private final Model model;

    public UpdateManager(Model model) {
        this.model = model;
    }

    public static final String CURRENT_VERSION = "1.4.9";
    public static final String RELEASES_LINK = "https://github.com/NobilityDeviant/Wcofun.com_Downloader/releases";
    protected final String githubLatest = "https://api.github.com/repos/NobilityDeviant/Wcofun.com_Downloader/releases/latest";
    public Update latestUpdate = null;

    private JsonObject apiSheet() {
        try {
            removeValidation();
            HttpsURLConnection urlConnection = (HttpsURLConnection) new URL(githubLatest).openConnection();
            urlConnection.setReadTimeout(20_000);
            urlConnection.setConnectTimeout(20_000);
            urlConnection.setInstanceFollowRedirects(true);
            urlConnection.setRequestProperty("user-agent", model.getRandomUserAgent());
            urlConnection.connect();
            BufferedReader in = new BufferedReader(new InputStreamReader(urlConnection.getInputStream()));
            StringBuilder stringBuilder = new StringBuilder();
            String inputLine;
            while ((inputLine = in.readLine()) != null) {
                stringBuilder.append(inputLine).append("\n");
            }
            in.close();
            urlConnection.disconnect();
            //return new JsonParser().parse(stringBuilder.toString()).getAsJsonObject();
            return JsonParser.parseString(stringBuilder.toString()).getAsJsonObject();
        } catch (Exception e) {
            System.out.println("Failed to get github api response. Error: " + e.getLocalizedMessage());
        }
        return null;
    }

    public Thread saveUpdateDetails() {
        return new Thread(() -> latestUpdate = latestUpdate());
    }

    public Update latestUpdate() {
        JsonObject json = apiSheet();
        if (json != null) {
            String version = json.get("tag_name").getAsString();
            String body = json.get("body").getAsString();
            JsonArray array = json.getAsJsonArray("assets");
            if (array != null) {
                for (JsonElement element : array) {
                    if (element.isJsonObject()) {
                        JsonObject o = element.getAsJsonObject();
                        if (o.has("browser_download_url")) {
                            String url = o.get("browser_download_url").getAsString();
                            if (url.endsWith(".jar")) {
                                return new Update(version, url, body);
                            }
                        }
                    }
                }
            }
        }
        return null;
    }

    private boolean isLatest(String latest) {
        if (latest == null || latest.equals(CURRENT_VERSION)) {
            return true;
        }
        try {
            String[] latestSplit = latest.split("\\.");
            String[] current = CURRENT_VERSION.split("\\.");
            if (Integer.parseInt(latestSplit[0]) > Integer.parseInt(current[0])) {
                return false;
            }
            if (Integer.parseInt(latestSplit[1]) > Integer.parseInt(current[1])) {
                return false;
            }
            if (Integer.parseInt(latestSplit[2]) > Integer.parseInt(current[2])) {
                return false;
            }
        } catch (Exception e) {
            e.printStackTrace();
            return true;
        }
        return false;
    }

    public void checkForUpdates(boolean prompt, boolean refresh) {
        new Thread(() -> {
            if (latestUpdate == null || refresh) {
                Thread thread = saveUpdateDetails();
                thread.start();
                try {
                    thread.join();
                } catch (InterruptedException e) {
                    System.out.println("Failed to check for updates. Error: " + e.getLocalizedMessage());
                    return;
                }
            }
            if (latestUpdate == null) {
                System.out.println("Failed to find latest update details.");
                return;
            }
            if (!model.settings().getString(Defaults.UPDATEVERSION).equalsIgnoreCase(latestUpdate.getVersion())) {
                model.settings().setBoolean(Defaults.DENIEDUPDATE, false);
                model.saveSettings();
            }

            if (!model.settings().getBoolean(Defaults.DENIEDUPDATE)) {
                model.settings().setString(Defaults.UPDATEVERSION, latestUpdate.getVersion());
                model.saveSettings();
                boolean latest = isLatest(latestUpdate.getVersion());
                if (latest && !prompt) {
                    return;
                }
                Platform.runLater(() ->
                        model.showUpdateConfirm((latest ? "Updated" : "Update Available") + " - ver. "
                                + latestUpdate.getVersion(), false, latest));
            }
        }).start();
    }

    private void removeValidation() {
        TrustManager[] trustAllCerts = new TrustManager[]{new X509TrustManager(){
            public X509Certificate[] getAcceptedIssuers(){return null;}
            @Override
            public void checkClientTrusted(X509Certificate[] arg0, String arg1) {
            }
            @Override
            public void checkServerTrusted(X509Certificate[] arg0, String arg1) {
            }
        }};
        try {
            SSLContext sc = SSLContext.getInstance("TLS");
            sc.init(null, trustAllCerts, new SecureRandom());
            HttpsURLConnection.setDefaultSSLSocketFactory(sc.getSocketFactory());
        } catch (Exception ignored) {}
    }
}

File: src/main/java/com/nobility/downloader/updates/UpdateManager.kt
package com.nobility.downloader.updates

import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.nobility.downloader.Model
import com.nobility.downloader.settings.Defaults
import com.nobility.downloader.utils.Resource
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.javafx.JavaFx
import kotlinx.coroutines.withContext
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.URL
import java.security.SecureRandom
import java.security.cert.X509Certificate
import javax.net.ssl.HttpsURLConnection
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManager
import javax.net.ssl.X509TrustManager

class UpdateManager(private val model: Model) {

    private val githubLatest = "https://api.github.com/repos/NobilityDeviant/Wcofun.com_Downloader/releases/latest"
    var latestUpdate: Update? = null

    private fun fetchLatestRelease(): Resource<JsonObject> {
        try {
            removeValidation()
            val urlConnection = URL(githubLatest).openConnection() as HttpsURLConnection
            urlConnection.readTimeout = 20000
            urlConnection.connectTimeout = 20000
            urlConnection.instanceFollowRedirects = true
            urlConnection.setRequestProperty("user-agent", model.randomUserAgent)
            urlConnection.connect()
            val `in` = BufferedReader(InputStreamReader(urlConnection.inputStream))
            val stringBuilder = StringBuilder()
            var inputLine: String?
            while (`in`.readLine().also { inputLine = it } != null) {
                stringBuilder.append(inputLine).append("\n")
            }
            `in`.close()
            urlConnection.disconnect()
            return Resource.Success(
                JsonParser.parseString(stringBuilder.toString()).asJsonObject
            )
        } catch (e: Exception) {
            return Resource.Error(e)
        }
    }

    private fun parseLatestRelease(): Resource<Update> {
        val result = fetchLatestRelease()
        if (result.data != null) {
            val json = result.data
            val version = json["tag_name"].asString
            val body = json["body"].asString
            val array = json.getAsJsonArray("assets")
            if (array != null) {
                for (element in array) {
                    if (element.isJsonObject) {
                        val o = element.asJsonObject
                        if (o.has("browser_download_url")) {
                            val url = o["browser_download_url"].asString
                            if (url.endsWith(".jar")) {
                                return Resource.Success(Update(version, url, body))
                            }
                        }
                    }
                }
            }
        } else if (result.message != null) {
            return Resource.Error(result.message)
        }
        return Resource.Error("Failed to parse github api response.")
    }

    suspend fun checkForUpdates(prompt: Boolean, refresh: Boolean) = withContext(Dispatchers.IO) {
        if (latestUpdate == null || refresh) {
            val result = parseLatestRelease()
            if (result.data != null) {
                latestUpdate = result.data
            } else {
                println(result.message)
            }
        }
        if (latestUpdate == null) {
            println("Failed to find latest update details. No error found.")
            return@withContext
        }
        if (!model.settings().stringSetting(Defaults.UPDATEVERSION)
                .equals(latestUpdate!!.version, ignoreCase = true)) {
            model.settings().setSetting(Defaults.DENIEDUPDATE, false)
        }
        if (!model.settings().booleanSetting(Defaults.DENIEDUPDATE)) {
            model.settings().setSetting(Defaults.UPDATEVERSION, latestUpdate!!.version)
            val latest = isLatest(latestUpdate!!.version)
            if (latest && !prompt) {
                return@withContext
            }
            withContext(Dispatchers.JavaFx) {
                model.showUpdateConfirm(
                    (if (latest) "Updated" else "Update Available") + " - ver. "
                            + latestUpdate!!.version, false, latest
                )
            }
        }
    }

    private fun isLatest(latest: String?): Boolean {
        if (latest == null || latest == CURRENT_VERSION) {
            return true
        }
        try {
            val latestSplit = latest.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            val current = CURRENT_VERSION.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            if (latestSplit[0].toInt() > current[0].toInt()) {
                return false
            }
            if (latestSplit[1].toInt() > current[1].toInt()) {
                return false
            }
            if (latestSplit[2].toInt() > current[2].toInt()) {
                return false
            }
        } catch (e: Exception) {
            e.printStackTrace()
            return true
        }
        return false
    }

    private fun removeValidation() {
        val trustAllCerts = arrayOf<TrustManager>(object : X509TrustManager {
            override fun checkClientTrusted(arg0: Array<X509Certificate>, arg1: String) {}
            override fun checkServerTrusted(arg0: Array<X509Certificate>, arg1: String) {}
            override fun getAcceptedIssuers(): Array<X509Certificate> {
                return emptyArray()
            }
        })
        try {
            val sc = SSLContext.getInstance("TLS")
            sc.init(null, trustAllCerts, SecureRandom())
            HttpsURLConnection.setDefaultSSLSocketFactory(sc.socketFactory)
        } catch (ignored: Exception) {
        }
    }

    companion object {
        const val CURRENT_VERSION = "1.4.9"
        const val RELEASES_LINK = "https://github.com/NobilityDeviant/Wcofun.com_Downloader/releases"
    }
}